precompiles/common/precompile.go

```diff
@@ -132,7 +135,24 @@ func (p Precompile) RunSetup(
 	// we need to consume the gas that was already used by the EVM
 	ctx.GasMeter().ConsumeGas(initialGas, "creating a new gas meter")
 
-	return ctx, stateDB, s, method, initialGas, args, nil
+	// add a snapshot of the current state before executing the precompile
+	// so that any out-of-gas errors during said execution are reverted correctly
+	if isTransaction(method.Name) {
+		if err := p.AddPrecompileSnapshot(stateDB, s); err != nil {
+			// if we have exceeded the limit of precompile calls, it will add the entry to the
+			// journal and return an error. then, we will revert that latest entry and hence
```

Comment:
for your comment, is adding a log entry and returning an error an atomic operation?
If adding the log entry succeeds but subsequent steps fail, will it result in inconsistent state?
Can the rollback handle such  situation correctly?

Reply:
Yes, it is an atomic operation. A unit test for this case exists in `wrapped_revert_test.go`.

- We take a snapshot of the state and dump everything from the statedDB to the cached context
- We add a journal entry
- Then, we check how many precompile call entries have been added
- If more entries than the threshold have been added, we return an error.
- This error deletes the latest entry and restores everything to the snapshot
- The cached context is reset to match the snapshot state
- `writeCache` is modified to emit the events and dump items to disk; it is called only at the end of the transaction